util/dvsim/SlurmLauncher.py

```diff
@@ -0,0 +1,137 @@
+        self.process = None
+        self.slurm_log_file = self.deploy.get_log_path() + '.slurm'
+
+    def _do_launch(self):
+        # replace the values in the shell's env vars if the keys match.
+        exports = os.environ.copy()
+        exports.update(self.deploy.exports)
+
+        # Clear the magic MAKEFLAGS variable from exports if necessary. This
+        # variable is used by recursive Make calls to pass variables from one
+        # level to the next. Here, self.cmd is a call to Make but it's
+        # logically a top-level invocation: we don't want to pollute the flow's
+        # Makefile with Make variables from any wrapper that called dvsim.
+        if 'MAKEFLAGS' in exports:
+            del exports['MAKEFLAGS']
+
+        self._dump_env_vars(exports)
+
+        try:
+            with open(self.slurm_log_file, 'w') as out_file:
+                out_file.write("[Executing]:\n{}\n\n".format(self.deploy.cmd))
+                out_file.flush()
+
+                # Add a command delimiter if necessary
+                slurm_setup_cmd = SLURM_SETUP_CMD
+                if slurm_setup_cmd != '' and not slurm_setup_cmd.endswith(';'):
+                    slurm_setup_cmd += ';'
+
+                # Encapsulate the run command with the slurm invocation
+                slurm_cmd = f'srun -p {SLURM_QUEUE} --mem={SLURM_MEM} --mincpus={SLURM_MINCPUS} ' \
+                            f'--time={SLURM_TIMEOUT} --cpus-per-task={SLURM_CPUS_PER_TASK} ' \
+                            f'bash -c "{slurm_setup_cmd} {self.deploy.cmd}"'
+
+                log.info(f'Executing slurm command: {slurm_cmd}')
+                self.process = subprocess.Popen(shlex.split(slurm_cmd),
+                                                bufsize=4096,
+                                                universal_newlines=True,
+                                                stdout=out_file,
+                                                stderr=out_file,
+                                                env=exports)
+        except subprocess.SubprocessError as e:
+            raise LauncherError(f'IO Error: {e}\nSee {self.deploy.get_log_path()}')
+        finally:
+            self._close_process()
+
+        self._link_odir("D")
+
+    def poll(self):
+        '''Check status of the running process
+
+        This returns 'D', 'P' or 'F'. If 'D', the job is still running. If 'P',
+        the job finished successfully. If 'F', the job finished with an error.
+
+        This function must only be called after running self.dispatch_cmd() and
+        must not be called again once it has returned 'P' or 'F'.
+        '''
+
+        assert self.process is not None
+        if self.process.poll() is None:
+            return 'D'
+
+        # Copy slurm job results to log file
+        if os.path.exists(self.slurm_log_file):
+            with open(self.slurm_log_file, 'r') as slurm_file:
+                lines = slurm_file.readlines()
```

Comment:
This is unefficient: you are reading the full file, breaking the string at each `\n` into multiple strings, and then writing this line by line. You might as well just do something like out_file.write(slurm_file.read()), or better yet, shutil.copyfileobj(slurm_file, out_file).

Reply:
Sounds good!